Automattic/woocommerce-services
# Add state machine to drive migration states

## Description
This PR replaces the task queue with a state machine to drive the migration process. This PR will remember where the last "stopped" state is should an error occur. This allows the user to continue where they left off. For example, if the user installed the plugin but failed to activate it, the next time they run the migration, it will start from "activating" instead of the beginning. 

![image](https://github.com/Automattic/woocommerce-services/assets/572862/d7f36d8c-9710-4975-92ec-10e99008c04d)

### Related issue(s)
N/A

### Steps to test
#### Happy path - test the full migration
1. Go to `woocommerce-services/classes/class-wc-connect-service-settings-store.php` and replace the following `is_eligible_for_migration` function. This function decides if the modal should popup or not
```
		public function is_eligible_for_migration() {
			WC_Connect_Options::delete_option( 'wcshipping_migration_state' );
			$migration_state = WC_Connect_Options::get_option( 'wcshipping_migration_state' );

			$migration_pending = WC_Connect_WCST_To_WCShipping_Migration_State_Enum::COMPLETED !== $migration_state;
			$migration_enabled = $this->service_schemas_store->is_wcship_wctax_migration_enabled();

			return $migration_pending && $migration_enabled;
		}
```
2. Make sure you have "WooCommerce Shipping" and "WooCommerce Tax" in your plugins folder. We can't test the "install" part because they aren't part of dotorg yet.
3. Go to any order page with WCS&T enabled, click the WCS&T "Create shipping label" red button,
4. Click "Update now"
![image](https://github.com/Automattic/woocommerce-services/assets/572862/6d0b1591-9b4f-4ab4-8807-9a88ce91f769)
5. It should install, activate Woo Shipping & Woo Tax, deactivate WCS&T, and redirect you to the plugins page.
6. [Optional] Check DB's `wp_options` table, `wc_connect_options` should have the value `..."wcshipping_migration_state";i:12;}` appended at the end. Key thing here is `12` is the "done state". 

#### Test 404 on install and activate, fix the issue, then continue
1. Repeat Step 1 to 3 above. Refresh the page so it deletes the `wcshipping_migration_state` option.
2. Now, comment out `WC_Connect_Options::delete_option( 'wcshipping_migration_state' );` from the happy path step 1). In this test, we want to keep `wcshipping_migration_state` so we can test the exceptions.

What we want to test here is to mimic a 404 errors downloading from dotorg. Or some API errors during activation. Here, we will test by tripping the circuit breaker at the `wc-admin/plugins/install` and `wc-admin/plugins/activate` API calls. The goal is to test if the migration can continue where it last failed.

3. `npm run start`. Open up `client/components/migration/migration-runner.js`. Append `2` to the API URL so they fail and throw `404`.  Change the following
- `fetch( getBaseURL() + 'wc-admin/plugins/install', {` to `fetch( getBaseURL() + 'wc-admin/plugins/install2', {`
- `fetch( getBaseURL() + 'wc-admin/plugins/activate', {` to `fetch( getBaseURL() + 'wc-admin/plugins/activate2', {`
4. Click the "Update now" button
5. You should see this in the console because `/install2` failed 
![image](https://github.com/Automattic/woocommerce-services/assets/572862/29650ec2-27f8-42e8-8cc5-62270a2179b3)
The `migration_state` value should be `5` because it failed at installation. 
5.1 [optional] Check DB `wp_options`, `wc_connect_options` should have value appended at the end `"wcshipping_migration_state";i:5;}`.
6. Now, fix the issue by changing `/install2` back to `install`. Refresh the page. `install` should work, and activate will now fail. The installation should pick up from the "install" step.
7. Confirm it starts from "install" and failed at "activate". The `migration_state` should be 7. 
![image](https://github.com/Automattic/woocommerce-services/assets/572862/a0c7ae9f-4723-45b5-bfba-d139ab8c8e00)
7.1. [optional] Check DB `wp_options`, `wc_connect_options` should have value appended at the end `"wcshipping_migration_state";i:7;}`.
8. Now, fix the `activate2` to `activate`. The migration should pick up from state `7` and go through the whole migration.
![image](https://github.com/Automattic/woocommerce-services/assets/572862/2c4ccf52-1518-415a-9f98-0918ba2c2d1c)
9. [optional] Check DB `wp_options`, `wc_connect_options` should have value appended at the end `"wcshipping_migration_state";i:12;}`.
10. You should be on the plugins page `http://localhost/wp-admin/plugins.php?plugin_status=all&paged=1&s`. Confirm the WCS&T is deactivated, and Woo Shipping and Woo Tax are enabled.
![image](https://github.com/Automattic/woocommerce-services/assets/572862/5517130f-ba29-4afd-8e83-4157450fcd8b)


### Checklist
